core/unit_test/TestGraph.hpp

```diff
@@ -349,14 +340,207 @@ TEST_F(TEST_CATEGORY_FIXTURE(graph), zero_work_reduce) {
-#endif
-#ifdef KOKKOS_ENABLE_HPX  // FIXME_HPX graph.submit() isn't properly enqueued
-  if constexpr (std::is_same_v<TEST_EXECSPACE, Kokkos::Experimental::HPX>)
-    Kokkos::fence();
 #endif
   graph.submit();
   Kokkos::deep_copy(ex, count_host, count);
   ex.fence();
   ASSERT_EQ(count_host(), 0);
 }
+
+// Ensure that an empty graph can be submitted.
+TEST_F(TEST_CATEGORY_FIXTURE(graph), empty_graph) {
+  auto graph = Kokkos::Experimental::create_graph(ex, [](auto) {});
+  graph.instantiate();
+  graph.submit(ex);
+  ex.fence();
+}
+
+template <typename ViewType, size_t TargetIndex, size_t NumIndices = 0>
+struct FetchValuesAndContribute {
+  static_assert(std::is_same_v<typename ViewType::value_type,
+                               typename ViewType::non_const_value_type>);
+
+  ViewType data;
+  typename ViewType::value_type value;
+  Kokkos::Array<size_t, NumIndices> indices{};
+
+  FetchValuesAndContribute(ViewType data_,
+                           std::integral_constant<size_t, TargetIndex>,
+                           typename ViewType::value_type value_)
+      : data(std::move(data_)), value(value_) {}
+
+  FetchValuesAndContribute(ViewType data_,
+                           Kokkos::Array<size_t, NumIndices> indices_,
+                           std::integral_constant<size_t, TargetIndex>,
+                           typename ViewType::value_type value_)
+      : data(std::move(data_)), value(value_), indices(std::move(indices_)) {}
+
+  template <typename T>
+  KOKKOS_FUNCTION void operator()(const T) const {
+    if constexpr (NumIndices > 0) {
+      for (size_t index = 0; index < indices.size(); ++index)
+        data(TargetIndex) += data(indices[index]);
+    }
+    data(TargetIndex) += value;
+  }
+};
+
+template <typename ViewType, size_t TargetIndex, size_t NumIndices>
+FetchValuesAndContribute(ViewType, const size_t (&)[NumIndices],
+                         std::integral_constant<size_t, TargetIndex>,
+                         typename ViewType::non_const_value_type)
+    -> FetchValuesAndContribute<ViewType, TargetIndex, NumIndices>;
+
+// Ensure that we can handle the simple diamond use case.
+//
+// topology     stream-based approach       graph-based
+//
+//   A          A(exec_0)                   Using the API to add nodes, no
+//  / \         fence(exec_0)               user-facing fence anymore because
+// B   C        B(exec_0)   C(exec_1)       we'd like to rely on the graph to
+//  \ /         fence(exec_1)               enforce dependencies.
+//   D          D(exec_0)
+TEST_F(TEST_CATEGORY_FIXTURE(graph), diamond) {
+  const auto execution_space_instances =
+      Kokkos::Experimental::partition_space(ex, 1, 1, 1, 1);
+
+  const auto exec_0 = execution_space_instances.at(0);
+  const auto exec_1 = execution_space_instances.at(1);
+  const auto exec_2 = execution_space_instances.at(2);
+  const auto exec_3 = execution_space_instances.at(3);
```

Comment:
When I see that I remember why I don't like the current interface and would like something that let the structured binding
```C++
auto const [exec_0, exec_1, exec_3] = partition_space(ex, 1, 1, 1);
```
(just a rant, no actionable item)

Reply:
In the past, I also expected that the `partition_space` should return a tuple-like container if weights are passed one-by-one, and a vector-like container if weights are passed with a vector.

I can open an issue for that, to gather people's feedback. I do think I would make sense.

Reply:
We will need `C++20` for structured bindings to be usable easily. For now you'd get something like (with `clang`):
```bash
error: reference to local binding 'a' declared in enclosing function 'f' 
```

So once `Kokkos` is on `c++20` we'll be able to do:
```c++
const auto [exec_0, ... exec_N] = Kokkos::Experimental::partition_space(ex, weight_0, ..., weight_N);
```

Reply:
@dalg24 OK to close ?